streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -169,51 +181,49 @@ private void initialize(final RestoringTasks active) {
+    private void startRestoration(final Set<TopicPartition> initialized,
                                   final RestoringTasks active) {
-        log.debug("Start restoring state stores from changelog topics {}", initialized.keySet());
+        log.debug("Start restoring state stores from changelog topics {}", initialized);
 
         final Set<TopicPartition> assignment = new HashSet<>(restoreConsumer.assignment());
-        assignment.addAll(initialized.keySet());
+        assignment.addAll(initialized);
         restoreConsumer.assign(assignment);
 
-        final List<StateRestorer> needsPositionUpdate = new ArrayList<>();
-        for (final StateRestorer restorer : initialized.values()) {
+        final Iterator<TopicPartition> iterator = initialized.iterator();
+
+        while (iterator.hasNext()) {
+            final TopicPartition partition = iterator.next();
+            final StateRestorer restorer = stateRestorers.get(partition);
             if (restorer.checkpoint() != StateRestorer.NO_CHECKPOINT) {
-                restoreConsumer.seek(restorer.partition(), restorer.checkpoint());
-                logRestoreOffsets(restorer.partition(),
-                                  restorer.checkpoint(),
-                                  endOffsets.get(restorer.partition()));
-                restorer.setStartingOffset(restoreConsumer.position(restorer.partition()));
+                restoreConsumer.seek(partition, restorer.checkpoint());
+                logRestoreOffsets(partition,
+                        restorer.checkpoint(),
+                        endOffsets.get(partition));
+                restorer.setStartingOffset(restoreConsumer.position(partition));
                 restorer.restoreStarted();
             } else {
-                final StreamTask task = active.restoringTaskFor(restorer.partition());
+                final StreamTask task = active.restoringTaskFor(partition);
+
+                restoreConsumer.seekToBeginning(Collections.singletonList(partition));
 
                 // If checkpoint does not exist it means the task was not shutdown gracefully before;
                 // and in this case if EOS is turned on we should wipe out the state and re-initialize the task
                 if (task.isEosEnabled()) {
                     log.info("No checkpoint found for task {} state store {} changelog {} with EOS turned on. " +
-                            "Reinitializing the task and restore its state from the beginning.", task.id, restorer.storeName(), restorer.partition());
-                    task.reinitializeStateStoresForPartitions(Collections.singleton(restorer.partition()));
+                            "Reinitializing the task and restore its state from the beginning.", task.id, restorer.storeName(), partition);
+
+                    needsInitializing.remove(partition);
+                    iterator.remove();
+                    restorer.setCheckpointOffset(restoreConsumer.position(partition));
```

Comment:
Above is `restoreConsumer.seekToBeginning(Collections.singletonList(partition))` and here for the `EOS` enabled case is `restorer.setCheckpointOffset(restoreConsumer.position(partition))` 

Should these two statements get flipped meaning `restoreConsumer.seekToBeginning(Collections.singletonList(partition))`  is in the `if(task.isEosEnabled())` branch?

Reply:
We should always reset to beginning for either case, the difference is that with EOS turned on we want to re-initialize the state store, whereas without EOS we can just override from beginning on whatever is in the current state dir.